### PR TITLE
[ENH] Add xLSTMTime model in v2

### LINKED ISSUE #2039: [ENH] Add v2 interface support for `xLSTMTime`
Referencing #1992 

**Is your feature request related to a problem? Please describe.**
The xLSTMTime model currently only supports the v1 interface. With the introduction of the v2 architecture using `TslibBaseModel` and `TslibDataModule`, xLSTMTime should be migrated to support the new interface for consistency with other v2 models and to leverage improvements in the v2 design.

**Describe the solution you'd like**
Implement `xLSTMTime_v2` with the following components:

1. **Model class** (`_xlstm_v2.py`): Extend `TslibBaseModel` with proper handling of:
   - Dynamic `input_size` and `output_size` derived from metadata
   - Both sLSTM and mLSTM variants
   - Proper dimension handling for BatchNorm and LSTM layers

2. **Package class** (`_xlstm_v2_pkg_v2.py`): Extend `Base_pkg` with:
   - Auto-computation of `input_size` (from continuous features) and `output_size` (prediction_length × n_targets)

3. **Integration**: Ensure compatibility with:
   - `TslibDataModule` for data handling
   - Standard v2 test suite (integration, checkpointing, prediction modes)

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

CC : @fkiraly

### PR DESCRIPTION
**Pull request by Faakhir30**
#### Reference Issues/PRs

Fixes #2039

#### What does this implement/fix? Explain your changes.

- Implements  `xLSTMTime` for v2
- adds package wrapper `xLSTMTime_pkg_v2` 

#### Did you add any tests for the change?

Nope

#### PR checklist

- [X] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.
- [X] Added/modified tests
- [X] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
  To run hooks independent of commit, execute `pre-commit run --all-files`

### CODE REVIEW

File: pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    _tags = {
+        "info:name": "xLSTMTime",
+        "info:compute": 3,
+        "info:y_type": ["numeric"],
+        "authors": ["Faakhir30"],
+        "capability:exogenous": True,
+        "capability:multivariate": False,
+        "capability:pred_int": True,
+        "capability:flexible_history_length": False,
```
Comment (Faakhir30) [inline pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: two changes from v1:
- `pred_int` should be `True`, right? because we can give quantile support in v2
- `capability:flexible_history_length` means history_length can be variable across `fit` calls? if thats the case then it should be False, refer #2338

File: pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            {"xlstm_type": "mlstm"},
+            {"num_layers": 2},
+            {"xlstm_type": "slstm", "input_projection_size": 32},
+            {
+                "xlstm_type": "mlstm",
+                "decomposition_kernel": 3,
+                "dropout": 0.2,
+            },
+            {
+                "loss": QuantileLoss(quantiles=[0.1, 0.5, 0.9]),
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: I remember v1 implementation didnt support quantile loss? I think you have added some improvements!?

**Review by phoeenniixx — CHANGES_REQUESTED** [review]
before a deep review - i have a question out of curiosity - is this implementation similar to what was present in v1? Or have you implemented this from scratch?

File: pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py (hunk 2 above)
Comment (Faakhir30) [reply pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: yes, as mentioned in above comment regarding changes from v1:

>two changes from v1:
>pred_int should be True, right? because we can give quantile support in v2
>capability:flexible_history_length means history_length can be variable across fit calls? if thats the case then it should be False, refer https://github.com/sktime/pytorch-forecasting/issues/2338

**Comment by Faakhir30** [comment]
> is this implementation similar to what was present in v1

yeah, apart from 2 things mentioned above, and ofcourse other minor things like:
- reading data from metadata
- conforming output format to v2 API in `forward`

File: pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py (hunk 2 above)
Comment (phoeenniixx) [reply pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: what i mean is, although v2 pipeline support he quantile loss, does the model implementation supports it?

**Comment by phoeenniixx** [comment]
> > is this implementation similar to what was present in v1
> 
> yeah, apart from 2 things mentioned above, and ofcourse other minor things like:
> 
> * reading data from metadata
> * conforming output format to v2 API in `forward`

Nice, happy to hear this!

File: pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py (hunk 2 above)
Comment (Faakhir30) [reply pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: hmm,
My current quantile support implementation has been adapted from `DLiner` and `Samformer`.
Those v2 models seems to define output dim as:
`output_dim = self.prediction_length * self.n_quantiles` 

https://github.com/sktime/pytorch-forecasting/blob/e7c7965cc08a52a566eb24661f31c06b49c88395/pytorch_forecasting/models/dlinear/_dlinear_v2.py#L244-L252

https://github.com/sktime/pytorch-forecasting/blob/e7c7965cc08a52a566eb24661f31c06b49c88395/pytorch_forecasting/models/samformer/_samformer_v2.py#L179-L186

So, idea currently implemented in all of these is to extend head dimention, as `QuantileLoss` expects predictions shaped (batch, horizon, n_quantiles) and then it applies pinball loss per quantile index / prediction_length.

Final statement:  the head supports it. the original core networks did not uniquely “know” about quantiles. Would you consider this as quantile support? 

head is part of the "model implementation", but personally, I'm unsure whether backpropagation in above senario will make models learn anything about quantiles. And even if it is intended/required.

Comment (phoeenniixx) [reply pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: > the original core networks did not uniquely “know” about quantiles. Would you consider this as quantile support?

Idts that is quantile support, I think it is more about if the network knows the quantiles? I think we should look at the original implementation (the paper and the repo), does that support it? I think it may be a possibility that I might have made this network simpler while implementing it in v1 as at that time I was still understanding the interface and I might have reduced the complexity... I dont remember exactly :)

Comment (Faakhir30) [reply pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: > I think we should look at the original implementation (the paper and the repo), does that support it?

- [xsltmtime paper](https://arxiv.org/pdf/2407.10240) dont mention `quantile` anywhere. I'll remove it from xLSTMTime.

- Should we also remove it from [dlinear](https://arxiv.org/pdf/2205.13504) and [samformer](https://arxiv.org/pdf/2402.10198)?? those papers also dont mention `quantile` anywhere.

Comment (phoeenniixx) [reply pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: > * Should we also remove it from [dlinear](https://arxiv.org/pdf/2205.13504) and [samformer](https://arxiv.org/pdf/2402.10198)?? those papers also dont mention `quantile` anywhere.

I agree, we should check if the models actually are aware of the quantiles or not

File: pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class xLSTMTime_pkg_v2(Base_pkg):
+    """xLSTMTime v2 package container."""
+
+    _tags = {
+        "info:name": "xLSTMTime",
+        "info:compute": 3,
+        "info:y_type": ["numeric"],
+        "authors": ["Faakhir30"],
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py]: we should add the authors from original implementation as well (and if the code is quite same as v1, then v1 authors as well)

File: pytorch_forecasting/models/xlstm/__init__.py [hunk 4]
```diff
@@ -2,5 +2,7 @@
 
 from pytorch_forecasting.models.xlstm._xlstm import xLSTMTime
 from pytorch_forecasting.models.xlstm._xlstm_pkg import xLSTMTime_pkg
+from pytorch_forecasting.models.xlstm._xlstm_pkg_v2 import xLSTMTime_pkg_v2
+from pytorch_forecasting.models.xlstm._xlstm_v2 import xLSTMTime as xLSTMTimeV2
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/__init__.py]: I think if we need to do this:
```python
from pytorch_forecasting.models.xlstm import xLSTMTimeV2
```
Then why not just rename the class itself? We can change these once we have a fully stable v2

File: pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    def get_datamodule_cls(cls):
+        """Get the underlying DataModule class."""
+        from pytorch_forecasting.data.data_module import (
+            EncoderDecoderTimeSeriesDataModule,
+        )
+
+        return EncoderDecoderTimeSeriesDataModule
+
+    @classmethod
+    def get_test_train_params(cls):
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/_xlstm_pkg_v2.py:38]: it would be good to add some losses in the params as well - they will be just point preditction ones now right?

File: pytorch_forecasting/models/xlstm/_xlstm_v2.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                "Error in xLSTMTime: xlstm_type must be either 'slstm' or 'mlstm'"
+            )
+
+        self.max_encoder_length = self.metadata["max_encoder_length"]
+        self.max_prediction_length = self.metadata["max_prediction_length"]
+        self.input_size = self.metadata["encoder_cont"] + 1
+
+        self.n_quantiles = 1
+        if hasattr(loss, "quantiles") and loss.quantiles is not None:
+            self.n_quantiles = len(loss.quantiles)
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/_xlstm_v2.py:114]: O thought we removed quantiles?

File: pytorch_forecasting/models/xlstm/_xlstm_v2.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        # output_size ~= forecast horizon; extend by n_quantiles when needed
+        self.output_size = self.max_prediction_length * self.n_quantiles
+
+        self.xlstm_type = xlstm_type
+        self.input_projection_size = input_projection_size or hidden_size
+
+        # clamp kernel so short test contexts still work (v1 assumed long history)
+        # kernel = min(decomposition_kernel, self.max_encoder_length)
+        # if kernel % 2 == 0:
+        #     kernel = max(1, kernel - 1)
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/_xlstm_v2.py]: can you please explain this block?

File: pytorch_forecasting/models/xlstm/_xlstm_v2.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        metadata: dict | None = None,
+        **kwargs,
+    ):
+        super().__init__(
+            loss=loss,
+            logging_metrics=logging_metrics,
+            optimizer=optimizer,
+            optimizer_params=optimizer_params,
+            lr_scheduler=lr_scheduler,
+            lr_scheduler_params=lr_scheduler_params,
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/_xlstm_v2.py:92]: a suggestion:
We should first collect all the params passed to `__init__` in `self` and then we should use them and not directly.

File: pytorch_forecasting/models/xlstm/_xlstm_v2.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    stabilized LSTM (sLSTM) or the matrix-memory mLSTM variant.
+
+    Parameters
+    ----------
+    loss : nn.Module
+        Loss (and evaluation metric) used during training.
+    hidden_size : int, default 32
+        Hidden size of the xLSTM network; also used by batch norm / LSTM internals.
+    xlstm_type : {"slstm", "mlstm"}, default "slstm"
+        Specifies which xLSTM variant to use:
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/_xlstm_v2.py:30]: we should add a blank line here so that the list is rendered correctly in the docs

File: pytorch_forecasting/models/xlstm/_xlstm_v2.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    lr_scheduler : str, optional
+        Learning rate scheduler name.
+    lr_scheduler_params : dict, optional
+        Parameters for the learning rate scheduler.
+    metadata : dict, optional
+        Metadata from the encoder-decoder datamodule. Used to derive
+        ``input_size`` (``encoder_cont + 1`` for ``target_past``) and
+        ``output_size`` (``max_prediction_length``).
+
+    Based on https://arxiv.org/pdf/2407.10240 and https://github.com/muslehal/xLSTMTime
```
Comment (phoeenniixx) [inline pytorch_forecasting/models/xlstm/_xlstm_v2.py]: this should be BEFORE `Parameters`

**Review by phoeenniixx — CHANGES_REQUESTED** [review]
Nice!
Just added few comments above

File: pytorch_forecasting/models/xlstm/_xlstm_v2.py (hunk 6 above)
Comment (Faakhir30) [reply pytorch_forecasting/models/xlstm/_xlstm_v2.py:114]: oh yeah, this is dead code, removed the `n_quantiles` in forward, updated, thanks

File: pytorch_forecasting/models/xlstm/__init__.py (hunk 4 above)
Comment (Faakhir30) [reply pytorch_forecasting/models/xlstm/__init__.py]: renamed to `xLSTMTime_v2` for now.
Can also be `xLSTMTimeModel` instead, or any other suggestion?

Renaming it to `xLSTMTimeV2` will require to make pkg name `xLSTMTimeV2_pkg_v2` (slightly awkward to have two v2 in a name) due to test `test_pkg_linkage`

File: pytorch_forecasting/models/xlstm/_xlstm_v2.py (hunk 7 above)
Comment (Faakhir30) [reply pytorch_forecasting/models/xlstm/_xlstm_v2.py]: forgot to remove commented code, was originally facing error on some test, removed.

File: pytorch_forecasting/models/xlstm/__init__.py (hunk 4 above)
Comment (phoeenniixx) [reply pytorch_forecasting/models/xlstm/__init__.py]: > Renaming it to `xLSTMTimeV2` will require to make pkg name `xLSTMTimeV2_pkg_v2` (slightly awkward to have two v2 in a name) due to test `test_pkg_linkage`

I thought this issue was solved by @PalakB09 ? Can you pls recheck and lmk. See #2080

Comment (phoeenniixx) [reply pytorch_forecasting/models/xlstm/__init__.py]: > renamed to `xLSTMTime_v2` for now.
> Can also be `xLSTMTimeModel` instead, or any other suggestion?

We can do this ofcourse, but think of the user who is using `ptf` and they wants to `xLSTMTime`, but as v2 is unstable, they cant use that in the production. So, to reduce the confusion, I think adding v2 at the end will be better so that the user knows which version of the model they are using!

Comment (Faakhir30) [reply pytorch_forecasting/models/xlstm/__init__.py]: > I thought this issue was solved by @PalakB09 ? Can you pls recheck and lmk. See https://github.com/sktime/pytorch-forecasting/pull/2080

It shows relaxation for `_v2` that I have already used by renaming model to `xLSTMTime_v2`.
It doesnt handles `V2` as a special case. 

`_v2` also seems fine, does bring the clarity of v2. 
If `xLSTMTimeV2` is strictly necessary, we need to allow `V2` also as a special case, lmk

**Review by phoeenniixx — APPROVED** [review]
Thanks!
